jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3740,11 +3844,19 @@ private static final void validateArgumentsOfIteratedLoop(MethodHandle iteratorH
 			
 			/* The signature of the passed-in loop body must be either (V, T, A...)V or (T, A...)void */
 			if (bodyReturnType != bodyParamTypes[0]) {
-				/*[MSG "K065D", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
-				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D", bodyReturnType.getName(), bodyParamTypes[0].getName())); //$NON-NLS-1$
+				MethodType correctType = bodyType.insertParameterTypes(0, bodyReturnType);
+				
+				/*[MSG "K065D2", "The return type of loop body does not match the type of the first argument: {0} != {1}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D2", bodyType.toString(), correctType.toString())); //$NON-NLS-1$
 			}
+			
+			bodyIterationVarLength = 2;
 		}
 		
+		/* If the iterator handle is non-null, it must have the return type java.util.Iterator or a subtype thereof.
+		 * If null, the internal parameter list must have at least one A type that is java.util.Iterator or a subtype.
+		 * Otherwise default A will be set to java.util.Iterator.
+		 */
 		if (null != iteratorHandle) {
```

Comment:
The following code should be optimized to make it understandable to reviewers.

Reply:
@ChengJin01 just pushed the optimized code that we discussed.